contrib/jms/README.md

```diff
@@ -0,0 +1,45 @@
+# JMS module
+
+This library provides Dataflow sources and sinkgs to make it possible to read
+and write on JMS brokers from Dataflow pipelines.
+
+It supports both JMS queues and topics, with unbounded or bounded `PCollections`.
+
+To use JmsIO, you have to:
+
+1. Create a JMS `ConnectionFactory` specific to the JMS broker you want to use (for instance, ActiveMQ, IBM MQ, ...)
+2. Specify the JMS destination (queue or topic) you want to use
+
+## Reading (consuming messages) with JmsIO
+
+The `JmsIO.Read` transform continuously consumes from the JMS broker and returns an unbounded `PCollection` of `Strings` that
+represent the messages. By default, each element in the resulting `PCollection` is encoded as a UTF-8 string.
+You can override the default encoding by using `withCoder` when you call JmsIO.Read.
+
+----
+PipelineOptions options = PipelineOptionsFactory.create();
+Pipeline p = Pipeline.create(options);
+
+// create JMS connection factory, for instance, with ActiveMQ
+javax.jms.ConnectionFactory connectionFactory = new ActiveMQConnectionFactory("tcp://localhost:61616");
+
+PCollection<String> data = p.apply(JmsIO.Read.named("ConsumeFromJMS").connectionFactory(connectionFactory).queue("my-queue"));
+----
+
+### Reading a bounded set of messages from JMS
+
+The `DirectPipelineRunner` and the batch mode of the Dataflow service do not support unbounded `PCollections`.
+To use JmsIO as source in these contexts, you need to supply a bound on the amount of messages to consume.
+
+You can specify the `.maxNumMessages` option to read a fixed maximum number of messages.
```

Comment:
I find it interesting that you separated out the above two sections. Is there a good reason for this?

I would have kept it simpler, and more unified, something like this:

## Reading from JMS
The `JmsIO.Read` transform produces an unbounded `PCollection` containing messages received from a JMS queue. By default, each element in the output collection is a `UTF-8`-encoded string. You can override the default encoding using `JmsIO.Read.withCoder`.

```java
Pipeline p = ...;

ConnectionFactory connectionFactory = new ActiveMQConnectionFactory("tcp://localhost:61616");

// Producing Strings by default
JmsIO.Read read = JmsIO.Read.connectionFactory(connectionFactory).queue("my-queue");
PCollection<String> data = p.apply("Read from ActiveMQ", read);

// Producing a byte[] using a different Coder
JmsIO.Read read = JmsIO.Read.connectionFactory(connectionFactory).queue("my-queue").withCoder(ByteArrayCoder.class);
```

`JmsIO` can also produce a *bounded* `PCollection` by limiting either the number of records returned (`Read.withMaxNumRecords`) or the time for which records are consumed (`Read.withMaxReadTime`). Note that if only a record limit is configured, your pipeline will not finish executing until the desired number of records is received.

```java
JmsIO.Read read = JmsIO.Read.connectionFactory(connectionFactory).queue("my-queue").withMaxNumRecords(5);
```

====

There should probably be a type variable on `JmsIO.Read`, like `JmsIO.Read<String>`, `JmsIO.Read<byte[]>`, but I haven't gotten to that code yet.